README.md
# joe-relics

This repository holds a collection of older versions of JOE, found from
various sources:

* `collection/usenet`: USENET messages and shar files from which several
  versions were released until 1.0.5
* `collection/cvs`: The full CVS respository that was used from version
  2.9 through 3.7
* `collection/downloads`: Archives downloaded from several sources
  (archives and otherwise) found across the internet. These would have
  originally been hosted on FTP sites in the mid-nineties.
* `collection/joe-hist`: The contents of the joe-hist archive released
  on Sourceforge, several years ago.
* `versions`: Contents of each complete release found in the above
  locations.

The Mercurial repository is still readily-available on
[Sourceforge](https://sourceforge.net/p/joe-editor/mercurial/ci/default/tree/)
and is not included here.

### gitport

Finally, the `gitport` directory contains source code and metadata necessary
to construct the git repository with full history and release tags. As this
is a one-shot deal, it is mostly a curiosity.

## Early JOE (< 2.9) analysis

Early JOE comprises 31 versions made in approximately 11 sprints (clusters
of releases made in quick succession).  Of these, almost half (15 versions)
survive, with representation here from every sprint.  Some releases from
this era were included in early Linux distributions, which accounted for
most of the editor's popularity:

* 1.0.0 in SLS 1.02
* 1.0.8 in Slackware 2.0.1, MCC 1.0, and some later SLS version
* 2.2 in Slackware 2.1 through 3.4
* 2.3 in RedHat 1.1 Mother's Day Release
* 2.8 in just about every distribution from 1995 and on

### Timeline

* **E** *1988-09-20*
* **0.x** series (7 versions in 2-3 parts)
	* First edition:
		* **J** *1991-08-21*
		* **0.1.0** *1991-09-03*
	* **0.1.1** *unknown*
	* Second edition:
		* **0.1.2** *1992-01-23*
		* **0.1.3** *unknown*
		* **0.1.4** *1992-01-30*
	* **0.1.5** *before 06-20, probably before 03-27*
* **1.0.x** series (14 versions in 4 parts)
	* 1992 series
		* **1.0.0** *1992-11-16*
			* SLS 1.02
		* **1.0.1** *1992-11-17*
		* **1.0.2** *unknown*
		* **1.0.3** *1992-11-18*
		* **1.0.4** *unknown*
		* **1.0.5** *1992-11-21*
		* **1.0.6** *1992-11-21*
		* **1.0.7** *1992-11-26*
		* **1.0.8** *1992-12-19*
			* Slackware 2.0.1
			* MCC 1.0
			* SLS
	* 1994 series
		* **1.0.9** *1994-04-24*
		* **1.0.10** *1994-08-09*
		* **1.0.11** *1994-08-10*
		* **1.0.12** *1994-08-18*
			* Slackware 2.3, possibly as early as 2.1
		* **1.0.13** *1994-08-24*
* **2.x** series (9 versions in 3 parts)
	* 1994 series: (3 versions in ~1 part)
		* **2.0** *probably 1994-09-22 or 1994-09-23*
		* **2.1** *1994-09-23*
		* **2.2** *1994-10-06*
			* Slackware 2.3 through 3.4
	* Christmas update:
		* **2.3** 1994-12-24 [1 versions, 1 part]
			* RedHat 1.1 (Mother's Day Release)
	* 1995 series: (5 versions in 1 part)
		* **2.4** *1995-01-07*
		* **2.5** *1995-01-08*
		* **2.6** *1995-01-10*
		* **2.7** *unknown*
		* **2.8** *1995-01-23* Final update
			* Everywhere

Some notable missing releases:
* 1.0.13: The final update of the 1.0.x series
* 2.0-2.1: The first posted changes post-refactor


gitport/NEWS.md
## Release Notes

[Back to README file](http://sourceforge.net/p/joe-editor/mercurial/ci/default/tree/README.md)

[Browse source](https://sourceforge.net/p/joe-editor/mercurial/ci/default/tree)

[Download source distribution](https://sourceforge.net/projects/joe-editor/files/)

[Build instructions](https://sourceforge.net/p/joe-editor/mercurial/ci/default/tree/INSTALL.md)

### JOE.next

* Bugs fixed

	* Fix a number of bugs involved with piping data into JOE:

		* Make piping into JOE binary clean.  For example, in: "echo -e 'a\0b' | joe" the NUL was missing.

		* Fix bug where 'New File' was displayed in "echo hi | joe -"

		* Fix bug where closing the copying process didn't cause the other end of the pipe feeding stdin from seeing SIGPIPE.

		* Fix bug where window resize will cause copying process to stop (needed to retry on EINTR).

	* Revert ^K Q behavior (by popular demand!)

### JOE 4.6

* Enhancements

	* Upgrade to Unicode 10.0.0.  Add configure environment variable
	  that allows you select Unicode 8.0.0, 9.0.0 or 10.0.0.

	* Install Gnome .desktop files for JOE.  This allows you to use JOE
	  in GUI "Open With" mouse actions.

	* Support strikeout and double-underline attributes for the few terminal
	  emulators that support them.  Use "stricken" and "dunderline" in
	  the syntax files, or \s and \z in help screens and status line.

* Bugs fixed

	* Fix bug where JOE would sometimes crash when editing shell
	  scripts.  This was due to an obscure bug in the syntax highlighter:
	  'reset' command (used for default state) was incorrectly messing
	  with stack.

	* -highlighter_context was missing from many file types, which
	  negated some improvements from the previous version.

	* Fixed signed char overflow with old style mouse events in large
	  windows

	* Fix bug where path restart (//) was being applied to block filter
	  command prompt.  Strange things would happen if you had adjacent
	  slashes in command arguments.

	* Allow ~ expansion but suppress path restart (//) in compiler error
	  parsing.

	* Restore default handling of SIGPIPE and SIGINT for shell commands. 
	  This fixes an issue where SIGPIPE doesn't terminate a process as
	  expected, for example by the head -n 10 in: ^K R !sh -c 'while :;
	  do echo y; done' | head -n 10.  This issue only occurred in
	  read/write to !, JOE already did the right thing for shell windows
	  and the filter region through shell command.

	* Improve screen update algorithm so that spaces at the ends of
	  lines are always emitted.  This allows them to be preserved when
	  cutting text with the mouse from a terminal emulator window.

	* Improve efficiency of screen update algorithm: JOE had been
	  resetting attributes such as background color before performing
	  cursor motions (probably as work around for bugs in old terminal
	  emulators).  This made screen update slow when there were many
	  attributes, as with syntax highlighting and color schemes.  It was
	  also repeatedly emitting ESC [ K.

	* Switch JOE to issue scrolling commands, even at high baud rates
	  (before this, JOE issued scrolling commands only at 19200 and
	  below because it used to be that simple screen refresh was faster
	  than scrolling in terminal emulators.  But this is no longer true
	  with complex screens involving color schemes, unicode and
	  highlighting).

* Windows version

	* Fix file access checks.  Inaccessible files due to ACL's weren't
	  correctly detected.  This was a regression in 4.5.

	* Check the correct (Windows) path in the home directory for color
	  schemes.  This was not re-implemented properly in 4.5.

### JOE 4.5

Before:

![Before](http://joe-editor.sourceforge.net/before.gif)

After:

![After](http://joe-editor.sourceforge.net/after.gif)

* Enhancements

	* New color scheme feature, which can be accessed with ^T S or
	  by setting the **-colors** option.

	* [24-bit color support](https://gist.github.com/XVilka/8346728)
	  can be enabled by setting the `COLORTERM` environment variable to
	  `truecolor` or `24bit`.

	* Several schemes included in-the-box:

		* gruvbox by Pavel Pertsev

		* ir_black by Todd Werth

		* molokai by Tomas Restrepo

		* solarized by Ethan Schoonover

		* wombat by Lars H. Nielsen

		* xoria by Dmitriy Y. Zotikov

		* zenburn by Jani Nurminen

	* The current line can be highlighted by pressing ^T U or by
	  enabling the **-hiline** option.

	* The gutter containing line numbers has a dynamic size based on the
	  length of the file, rather than a fixed size of 10.

	* Updated all language syntax files to use comment_todo and
	  string/comment contexts where appropriate.

	* Now pass character which invoked a macro to each macro step and
	  call.  If a macro step happens to be the 'type' command, the
	  character which invoked the macro will be typed in.  For example,
	  this macro will type three 'X's.  Before this change you got three
	  NULs.

		type,type,type   X

* Bugs fixed

	* Fix exsave: (^K ^X) should close file when a block is present in
	  the window, and the file is unmodified (regression from ^C change
	  in 4.2).

	* Fix regex assertions: they were not working because the character
	  before the search position was not being loaded.

	* For jmacs: ^Q^J now again inserts \n in the string replace prompt. 
	  This broke beginning with JOE 4.0.

	* Don't try to open files for writing to check file access.  This
	  behaves better on unionfs mounts.

	* Fix stray blocks created after find/replace.

* Windows version

	* Fix inability to set indent step to 1 from menu.

	* Add support for math functions, fix engineering display.

	* Fix backslash escapes in find/replace.

	* Support italicized text.

### JOE 4.4

* Enhancements

* Bugs fixed

	* Fix segfault due to buffer overrun.  This happens if a line
	  with many backslashes appears in the status line context display.

	* Fix jmacs: ^X ^F and ^X ^B were not working

	* Build fixes for Solaris


versions/joetty/cap.h
/* TERMCAP/TERMINFO header file
   Copyright (C) 1991 Joseph H. Allen

This file is part of JOE (Joe's Own Editor)

JOE is free software; you can redistribute it and/or modify it under the terms
of the GNU General Public License as published by the Free Software
Foundation; either version 1, or (at your option) any later version.  

JOE is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
A PARTICULAR PURPOSE.  See the GNU General Public License for more details.  

You should have received a copy of the GNU General Public License
along with JOE; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */

#define CAPEP 0			/* Flag capabilities */
#define CAPHD 1
#define CAPLC 2
#define CAPMT 3
#define CAPNL 4
#define CAPOP 5
#define CAPUC 6
#define CAPam 7
#define CAPbs 8
#define CAPbw 9
#define CAPca 10
#define CAPda 11
#define CAPdb 12
#define CAPeo 13
#define CAPgn 14
#define CAPhc 15
#define CAPhs 16
#define CAPhz 17
#define CAPin 18
#define CAPkm 19
#define CAPmi 20
#define CAPms 21
#define CAPmt 22
#define CAPnc 23
#define CAPns 24
#define CAPos 25
#define CAPpt 26
#define CAPul 27
#define CAPxb 28
#define CAPxn 29
#define CAPxo 30
#define CAPxr 31
#define CAPxs 32
#define CAPxt 33
#define CAPxv 34
#define CAPxx 35
#define NFLGS 36		/* Number of flag capabilities */
#define CAPco 0			/* Numeric capabilities */
#define CAPdB 1
#define CAPdC 2
#define CAPdF 3
#define CAPdN 4
#define CAPdT 5
#define CAPdV 6
#define CAPit 7
#define CAPkn 8
#define CAPli 9
#define CAPlm 10
#define CAPpb 11
#define CAPsg 12
#define CAPss 13
#define CAPtw 14
#define CAPug 15
#define CAPvt 16
#define CAPws 17
#define NNUMS 18		/* Number of numeric capabilities */
#define CAPAL 0			/* String capabilities */
#define CAPCC 1               
#define CAPCM 2
#define CAPDC 3
#define CAPDL 4
#define CAPDO 5
#define CAPIC 6
#define CAPK1 7
#define CAPK2 8
#define CAPK3 9
#define CAPK4 10
#define CAPK5 11
#define CAPLE 12
#define CAPRI 13
#define CAPSF 14
#define CAPSR 15
#define CAPUP 16
#define CAPae 17
#define CAPal 18
#define CAPas 19
#define CAPbc 20
#define CAPbl 21
#define CAPbt 22
#define CAPcd 23
#define CAPce 24
#define CAPch 25
#define CAPcl 26
#define CAPcm 27
#define CAPcr 28
#define CAPcs 29
#define CAPct 30
#define CAPcv 31
#define CAPdc 32
#define CAPdl 33
#define CAPdm 34
#define CAPdo 35
#define CAPds 36
#define CAPec 37
#define CAPed 38
#define CAPei 39
#define CAPes 40
#define CAPff 41
#define CAPfs 42
#define CAPge 43
#define CAPgo 44
#define CAPgt 45
#define CAPhd 46
#define CAPho 47
#define CAPhu 48
#define CAPi1 49
#define CAPi2 50
#define CAPi3 51
#define CAPiP 52
#define CAPic 53
#define CAPif 54
#define CAPim 55
#define CAPip 56
#define CAPis 57
#define CAPk0 58
#define CAPk1 59
#define CAPk2 60
#define CAPk3 61
#define CAPk4 62
#define CAPk5 63
#define CAPk6 64
#define CAPk7 65
#define CAPk8 66
#define CAPk9 67
#define CAPkA 68
#define CAPkC 69
#define CAPkD 70
#define CAPkE 71
#define CAPkF 72
#define CAPkH 73
#define CAPkI 74
#define CAPkL 75
#define CAPkM 76
#define CAPkN 77
#define CAPkP 78
#define CAPkR 79
#define CAPkS 80
#define CAPkT 81
#define CAPka 82
#define CAPkb 83
#define CAPkd 84
#define CAPke 85
#define CAPkh 86
#define CAPkl 87
#define CAPko 88
#define CAPkr 89
#define CAPks 90
#define CAPkt 91
#define CAPku 92
#define CAPl0 93
#define CAPl1 94
#define CAPl2 95
#define CAPl3 96
#define CAPl4 97
#define CAPl5 98
#define CAPl6 99
#define CAPl7 100
#define CAPl8 101
#define CAPl9 102
#define CAPle 103
#define CAPll 104
#define CAPma 105
#define CAPmb 106
#define CAPmd 107
#define CAPme 108
#define CAPmh 109
#define CAPmk 110
#define CAPml 111
#define CAPmm 112
#define CAPmo 113
#define CAPmp 114
#define CAPmr 115
#define CAPmu 116
#define CAPnd 117
#define CAPnl 118
#define CAPnw 119
#define CAPpO 120
#define CAPpc 121
#define CAPpf 122
#define CAPpk 123
#define CAPpl 124
#define CAPpo 125
#define CAPps 126
#define CAPpx 127
#define CAPr1 128
#define CAPr2 129
#define CAPr3 130
#define CAPrc 131
#define CAPrf 132
#define CAPri 133
#define CAPrp 134
#define CAPrs 135
#define CAPsa 136
#define CAPsc 137
#define CAPse 138
#define CAPsf 139
#define CAPso 140
#define CAPsr 141
#define CAPst 142
#define CAPta 143
#define CAPtc 144
#define CAPte 145
#define CAPti 146
#define CAPts 147
#define CAPuc 148
#define CAPue 149
#define CAPup 150
#define CAPus 151
#define CAPvb 152
#define CAPve 153
#define CAPvi 154
#define CAPvs 155
#define CAPwi 156
#define NSTRS 157		/* Number of string capabilities */

#define ENTRYLEN 1024
#define TERMPATH ".termcap /etc/termcap"

struct cap
 {
 I flgs[NFLGS];
 I nums[NNUMS];
 C *strs[NSTRS];
 C tbuf[ENTRYLEN];
 };

/* CAP *getcap(C *s);  Get CAP entry for terminal type name given in s.
   If s is zero, uses TERM environment variable.  Returns 0 if there was any
   error. */
CAP *getcap();

/* V texec(C *str,V (*out)(V *ptr,C c),V *outptr,
              C *pad,I div,I l,I x,I y);

   Execute a termcap string capability.
   
   'str' is the string to execute
   
   'out' is a function with each character to send to the terminal

   'outptr' is value passed as first argument to 'out'
   
   'pad' is a string containing the pad character to use.  If 'pad'==0,
   zero is used as the pad character.
   
   'div' is the number of 10ths of milliseconds it takes to send each
   character out at the current baud rate.  This is used to determine the
   number of padding characters to send.

   'l' is the number of lines effected by this string.  For example, if you
   use the clear to end of screen capability, the number of lines between
   the current cursor position and the end of the screen should be
   given here.

   'x' and 'y' are the arguments for the string.
*/
V texec();


/* I tcost(C *str,C *pad,I div,I l,I x,I y);
   Return cost in number of characters which need to be sent required
   to execute a termcap string capability.
   
   'str' is the string to execute
   
   'pad' is a string containing the pad character to use.  If 'pad'==0,
   zero is used as the pad character.
   
   'div' is the number of 10ths of milliseconds it takes to send each
   character out at the current baud rate.  This is used to determine the
   amount of padding characters to send.

   'l' is the number of lines effected by this string.  I.E., if you
   use the clear to end of screen capability, the number of lines between
   the current cursor position and the end of the screen should be
   given here.

   'x' and 'y' are the arguments for the string.
*/
I tcost();

/* C *tcompile(C *str,I x,I y,I *l);

   Compile a string capability.  Returns a pointer to a static area of memory
   containing the zero-terminated compiled string.  You must copy this string
   elsewhere before calling tcompile again.

   The length of the string is placed in the integer at l.
*/
C *tcompile();

/* V tshow(C *str);
   Print a compiled string on stdout:  Shows control characters in a nice way
 */
V tshow();

/* V rmcap(CAP *);  Eliminate a termcap entry returned by getcap */
V rmcap();
